custom_components/resmed_myair/client/rest_client.py

```diff
@@ -0,0 +1,704 @@
+            okta_url=self._region_config["okta_url"],
+            auth_server_id=self._region_config["auth_server_id"],
+        )
+        _LOGGER.debug(f"[get_initial_dt] initial_dt_url: {initial_dt_url}")
+        _LOGGER.debug(
+            f"[get_initial_dt] headers: {async_redact_data(self._json_headers, KEYS_TO_REDACT)}"
+        )
+
+        async with self._session.get(
+            initial_dt_url,
+            headers=self._json_headers,
+            raise_for_status=False, # This will likely return a 400 which is ok. Just need the device token.
+            allow_redirects=False,
+        ) as initial_dt_res:
+            _LOGGER.debug(f"[get_initial_dt] initial_dt_res: {initial_dt_res}")
+
+        await self._extract_and_update_cookies(initial_dt_res.headers.getall('set-cookie', []))
+
+    async def _is_access_token_active(self) -> bool:
+        introspect_url: str = OAUTH_URLS["introspect_url"].format(
+            okta_url=self._region_config["okta_url"],
+            auth_server_id=self._region_config["auth_server_id"],
+        )
+
+        headers: dict[str, Any] = {
+            "Accept": "application/json",
+            "Content-Type": "application/x-www-form-urlencoded",
+        }
+
+        introspect_query: dict[str, Any] = {
+            "client_id": self._region_config["authorize_client_id"],
+            "token_type_hint": "access_token",
+            "token": self._access_token,
+        }
+        _LOGGER.debug(f"[is_access_token_active] introspect_url: {introspect_url}")
+        _LOGGER.debug(
+            f"[is_access_token_active] headers: {async_redact_data(headers, KEYS_TO_REDACT)}"
+        )
+        _LOGGER.debug(
+            f"[is_access_token_active] introspect_query: {async_redact_data(introspect_query, KEYS_TO_REDACT)}"
+        )
+
+        async with self._session.post(
+            introspect_url, headers=headers, data=introspect_query, cookies=self._cookies
+        ) as introspect_res:
+            _LOGGER.debug(f"[is_access_token_active] introspect_res: {introspect_res}")
+            introspect_dict: dict[str, Any] = await introspect_res.json()
+            _LOGGER.debug(
+                f"[is_access_token_active] introspect_dict: {async_redact_data(introspect_dict, KEYS_TO_REDACT)}"
+            )
+            await self._resmed_response_error_check(
+                "introspect_query", introspect_res, introspect_dict
+            )
+        if introspect_dict.get("active", None) is True:
+            _LOGGER.info("Existing Access Token is already active. Reusing")
+            return True
+        return False
+
+    async def _resmed_response_error_check(
+        self, step: str, response: ClientResponse, resp_dict: dict, initial: bool | None = False
+    ) -> None:
+        if "errors" in resp_dict:
+            try:
+                error_message: str = f"{resp_dict['errors'][0]['errorInfo']['errorType']}: {resp_dict['errors'][0]['errorInfo']['errorCode']}"
```

Comment:
there's a risk that we get an error that is not well formed here

Reply:
I'm not sure I understand, sorry.